lib/java/ydoc-server/src/test/java/org/enso/ydoc/YdocTest.java

```diff
@@ -45,39 +67,149 @@ public void setup() {
   @After
   public void tearDown() throws Exception {
     ls.stop();
-    webServerExecutor.shutdownNow();
+    webServerExecutor.shutdown();
+    var stopped = webServerExecutor.awaitTermination(3, TimeUnit.SECONDS);
+    if (!stopped) {
+      var pending = webServerExecutor.shutdownNow();
+      log.error("Executor pending [{}] tasks: [{}].", pending.size(), pending);
+    }
     ydoc.close();
   }
 
   @Test
-  public void connect() throws Exception {
+  public void initialize() throws Exception {
     var queue = new LinkedBlockingQueue<BufferData>();
-    var url = YDOC_URL + "?ls=" + WEB_SERVER_URL;
 
     ydoc.start();
 
     var ws = WsClient.builder().build();
-    ws.connect(url, new TestWsListener(queue));
+    ws.connect(ydocUrl("index"), new DashboardConnection(queue));
 
-    var msg = queue.take();
-    Assert.assertArrayEquals(new byte[] {0, 0, 1, 0}, msg.readBytes());
+    var ok1 = queue.take();
+    Assert.assertTrue(ok1.debugDataHex(), BufferDataUtil.isOk(ok1));
+
+    var buffer = queue.take();
+    var uuid = BufferDataUtil.readUUID(buffer);
+    WsClient.builder().build().connect(ydocUrl(uuid.toString()), new DashboardConnection(queue));
+
+    var ok2 = queue.take();
+    Assert.assertTrue(ok2.debugDataHex(), BufferDataUtil.isOk(ok2));
+  }
+
+  private static final class BufferDataUtil {
+
+    private static final int UUID_BYTES = 36;
+    private static final int SUFFIX_BYTES = 3;
+
+    private static boolean isOk(BufferData data) {
+      return data.readInt16() == 0;
+    }
+
+    private static UUID readUUID(BufferData data) {
+      try {
+        data.skip(data.available() - UUID_BYTES - SUFFIX_BYTES);
+        var uuidString = data.readString(36);
+        return UUID.fromString(uuidString);
+      } catch (Exception e) {
+        log.error("Failed to read UUID of\n{}", data.debugDataHex());
+        throw e;
+      }
+    }
   }
 
-  private static final class TestWsListener implements WsListener {
+  private static final class DashboardConnection implements WsListener {
+
+    private static final Logger log = LoggerFactory.getLogger(DashboardConnection.class);
 
     private final BlockingQueue<BufferData> messages;
 
-    TestWsListener(BlockingQueue<BufferData> messages) {
+    private DashboardConnection(BlockingQueue<BufferData> messages) {
       this.messages = messages;
     }
 
     @Override
     public void onMessage(WsSession session, BufferData buffer, boolean last) {
+      log.debug("Got message\n{}", buffer.debugDataHex());
+
       messages.add(buffer);
     }
+
+    @Override
+    public void onMessage(WsSession session, String text, boolean last) {
+      log.error("Got unexpected message [{}].", text);
+    }
   }
 
-  private static final class WebServerWsListener implements WsListener {
-    WebServerWsListener() {}
+  private static final class LanguageServerConnection implements WsListener {
```

Comment:
Is this a mock server, mocking language server protocol. The real Ydoc code is connecting to it and we are verifying it connects and exchanges few messages, rigth?

Reply:
Correct